### ISSUE TITLE
Start.lsl and ServiceReceiver.lsl appear to have memory leaks

### ISSUE THREAD

**Issue by emmamei** [emmamei/cdkey#151]
Examine code for memory leaks; `ServiceReceiver.lsl` in particular appears to have memory leak.

**Comment by silkymesmeriser** [comment]
Mono uses about 25kB to store all the settings and ServiceReciever is still currently the internal database script is that what you are referring to?  Run time memory use for data storage and memory leaks are different.

**Comment by emmamei** [comment]
True. It seems to go up whenever a post is made to the database by `ServiceRequester`. It has also ran out of memory multiple times recently.

`Start.lsl` has also run out of memory recently, but much less often.

This is very interesting considering starting free space is well over 10kb.

**Comment by silkymesmeriser** [comment]
That there would be an increase in ServiceReceiver after a request is a logical consequence of there being a request.  Flow diagram:

ServiceRequester -> Request = Sent -> Services -> Reply = Received -> ServicesReceiver

So is it unexpected that there would be an increase in memory usage until the local variables that are part of the receiving and processing of a reply are cleaned up by the next garbage collector cycle not really.

As for Start, that it would have less memory free after finishing loading as it does initially also not really unexpected for a script which along with three arbitrary length user defined string variables and various short strings created at run time like the RLV API version variable it also has a string that contains the processed contents of the DataAppearance notecard which it reads and uses to restore such as the glow settings after the key is set invisible and then visible again though with any luck at some point LL will add the new materials features to llSet(Link)?PrimitiveParams(Fast)? in which case it wont be required I'm pretty sure when I tested that alpha masked surfaces beyond threshold go entirely transparent as if they literally do not exist including their glow effects however the same is not the case for alpha blended surfaces.  Unfortunately alpha masking cannot be set by script and any change to the alpha value results in the alpha mode being reset to alpha blending so for now the only option is to set the glow parameters to 0 destroying the original value in the process meaning we need the originals stored somewhere to restore it.

**Comment by silkymesmeriser** [comment]
This all said it would be a waste of development time to invest a significant amount of resources trying to invest much effort in trying to eliminate this problem with the scripts as they stand if we still intend to reduce the issues like the occasional corruption of lists like the controllers list and similar that comes from the undifferentiated variable length pipe separated lists of unknown intended length etc making them almost impossible to validate effectively with almost becoming infeasible if you eliminate the availability of regular expressions.

So this just adds to the growing list of reasons why it's important to get things all sorted in this area added to the current goal of being able to save a key state in a notecard and use it to restore settings as a backup system or as a poor man's substitute for a dynamic read/write database like the HTTPdb which is entirely feasable just so long as the database stuff is seperated out from the rest of service receiver, having to have multiple copies of the validation stuff in different scripts as external data comes in and gets processed randomly by scripts just about all over the map is also stupidly illogical repetitive and inefficient.

The solution of course is to actually have a data source agnostic script function handle all of this sort of thing in a logical clear and consistent way regardless where a particular bit of external data arrives from, this is something I've already thought a design plan through in my head for and plan to implement it alongside enabling and using JSON to the fullest to describe the data being posted to the database and returned, I picture the saved state notecards using the same format also.  I'm planning to develop both together along with the revamping of the code on the service side both to make it easier to read and understand, eliminate the dump everything in one file situation which made sense when the initial services were put together to try it out as all three services together were so simple and tiny, since the initial demonstration and getting then it getting the support of the rest of the team of course it has slowly evolved into a more robust but of course that then leads to it having a bit more code than read $_POST array, SQL escape input, connect to DB, run queries and it is becoming slowly more like you would expect the back end of a production application to be.  That combined with the fact that without making things really complicated with conditionals galore it's not that easy to make the enhanced JSON based API entirely backwards compatible with the old x-forms based concept that doesn't have the same capabilities, a JSON based API is better planned out and coded to be a JSON based API.

The work I started doing on the improved services end however I've been working on doing alongside the existing one the intent being that we can do a phased implementation on it starting out parallel running the two while we get everything implemented and tested to our satisfaction before going any further on it.  The entire point of course with the new API however being to develop a robust, production ready, stable and consistent structure for the way data is encoded and presented strongly supports dropping the code, hack till working, try and document development method and going with the analyse and define problem, design solution, document then implement method pretty sure W3C didn't start work on HTML5 hacking together an implementation then figuring out how to document it and very much doubt that is the order of things when the IETF develops a new open protocol standard either lol.

Should probably explain some more about the thoughts I already had on it though rather than dumping it all in one pile will probably make a meta issue and link the parts to it.

**Comment by emmamei** [comment]
There's a lot there, and it's a bit hard to follow. This bug is about how to keep the `ServicesReceiver.lsl` and `Start.lsl` scripts from running out of memory.

The reason for the failure of these scripts is irrelevant; running out of memory is not a valid response of a script to more data. `Start.lsl` for one should be largely static after load, with the usual back and forth of code within it.

`ServicesReceiver.lsl` shouldn't have a continuous record of all posts to the database I would think, but rather only the most current value of the affected data.

As I said, the reason for the script failure is irrelevant really; the important thing is to prevent the failure from occurring.

**Comment by emmamei** [comment]
Found a goodly amount of variable declarations inside of loops in the `doVisibility` code in `Start.lsl`; fixed. We'll see if that lowers the memory usage; most likely it will.

Most declarations were moved out of the loops; one was moved out of the inner loop to the outer loop, making it run less often.

If a declaration is inside a loop, then every time through the loop the variable is not only declared but "undeclared" at the end - causing an allocation of memory for a variable, followed by a possible release (or not) of memory.

**Comment by emmamei** [comment]
This may be fixed in `Start.lsl`; the `appearanceData` variable was read from the notecard but never reset - so a reread of the notecard resulted in more memory being used.

**Comment by emmamei** [comment]
No problems have been seen in `Start.lsl`; this part can be considered complete. `ServiceReceiver.lsl` continues to run out of memory.

**Comment by emmamei** [comment]
This could very well be related to the variable `storedConfigs` - the debug output reports it takes 45k to store.

**Comment by emmamei** [comment]
Problems with ServiceReceiver are moot since the backend has gone away. Closing.
